src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -0,0 +1,472 @@
+        //private vs non-primary alignments
+        private enum ReadsPartition {
+            PRIMARY, NOT_PRIMARY
+        }
+
+        /**
+         *  The header for all reads processed.
+        */
+        private final PCollectionView<SAMFileHeader> header;
+
+        public MarkDuplicatesDataflowTransform( final PCollectionView<SAMFileHeader> header ) {
+            this.header = header;
+        }
+
+        @Override
+        public PCollection<Read> apply( final PCollection<Read> preads ) {
+            final PCollectionList<Read> readsPartitioned = partitionReadsByPrimaryNonPrimaryAlignment(preads);
+
+            final PCollection<Read> fragments = readsPartitioned.get(ReadsPartition.PRIMARY.ordinal());
+            final PCollection<Read> fragmentsTransformed = transformFragments(header, fragments);
+
+            final PCollection<Read> pairs = readsPartitioned.get(ReadsPartition.PRIMARY.ordinal());
+            final PCollection<Read> pairsTransformed = transformReads(header, pairs);
+
+            //no work on those
+            final PCollection<Read> not_primary = readsPartitioned.get(ReadsPartition.NOT_PRIMARY.ordinal());
+
+            return PCollectionList.of(fragmentsTransformed).and(pairsTransformed).and(not_primary).apply(Flatten.<Read>pCollections());
+        }
+
+        private static PCollectionList<Read> partitionReadsByPrimaryNonPrimaryAlignment(PCollection<Read> preads) {
+            return preads.apply(Partition.of(2, new PartitionFn<Read>() {
+                private static final long serialVersionUID = 1l;
+
+                @Override
+                    public int partitionFor(final Read read, final int n) {
+                        return isPrimaryAlignment(read) ? ReadsPartition.PRIMARY.ordinal() : ReadsPartition.NOT_PRIMARY.ordinal();
+                    }
+                }));
+        }
+        /**
+         * (1) Reads are grouped by read group and read name.
+         * (2) Reads are then paired together as follows:
+         *   (a) The remaining reads (one per fragment key) are coordinate-sorted and paired
+         *       consecutively together and emitted.
+         *   (b) If a read is leftover, it is emitted, unmodified, as an unpaired end.
+         * (3) Paired ends are grouped by a similar key as above but using both reads.
+         * (4) Any unpaired end is emitted, unmodified.
+         * (5) The remained paired ends are scored and all but the highest scoring are marked as
+         *     duplicates. Both reads in the pair are emitted.
+         */
+        private PCollection<Read> transformReads(final PCollectionView<SAMFileHeader> headerPcolView, final PCollection<Read> pairs) {
+            final PCollection<KV<String, Iterable<Read>>> pairsGroupedByKey = groupPairsByKey(pairs, headerPcolView);
+            final PTransform<PCollection<? extends KV<String, Iterable<Read>>>, PCollection<KV<String, PairedEnds>>> markGroupedDuplicatePairs = markGroupedDuplicatePairs(headerPcolView);
```

Comment:
After a short fight with the type system I gave up. It wants the wildcards. Happy to fix if someone knows how to get rid of the silly wildcards.